### ISSUE TITLE
Allow setnames to include old columns not in x

### ISSUE THREAD

**Issue by MaxGhenis** [Rdatatable/data.table#1099]
A `allow.absent.cols` option for `setnames` would facilitate cases when the user wants to apply a column mapping without necessarily knowing whether all columns will exist. I'm currently using the following workaround, from [my SO self-answer](http://stackoverflow.com/a/29380605/1840471):

```r
Setnames <- function(x, old, new, allow.absent.cols=F) {
  if (!allow.absent.cols) {
    setnames(x, old, new)
  } else {
    old.intersect <- intersect(old, names(x))
    common.indices <- old %in% old.intersect
    new.intersect <- new[common.indices]
    setnames(x, old.intersect, new.intersect)
  }
}
```

**Comment by eantonya** [comment]
I'm not convinced this is a good idea, and if implemented it would need to come with a plethora of warnings.

**Comment by arunsrinivasan** [comment]
This doesn't seem like a good idea to me as well. If you need to, then you can always use column numbers instead of names. 

Unless there's a strong reasonable explanation for this, we should close this.

**Comment by MaxGhenis** [comment]
My use case involves looping through lots of raw datasets, and applying a single column name mapping to clean each in a consistent way. Something like

```
l <- list(...)  # List of raw data.tables, which don't all have the same set of 
raw.names <- c(...)  # Names showing up in raw data
clean.names <- c(...)  # Clean names
for (dt in l) setnames(dt, raw.names, clean.names)
```

I agree a warning for each skipped column makes sense.

**Comment by arunsrinivasan** [comment]
You just have to add 1 more line...

```
for (dt in l) {
    ix = match(names(dt), raw.names, 0L)
    setnames(dt, raw.names[ix], clean.names[ix])
}
```

**Comment by MaxGhenis** [comment]
Thanks, that's cleaner than my `intersect` / `%in%` approach above. Listed as an [answer](http://stackoverflow.com/a/29401988/1840471) on the SO question. I'm fine closing if this need is unusual.

**Comment by arunsrinivasan** [comment]
Glad that helped. :+1:

**Comment by MusYD** [comment]
I think that a good feature to add to this function would be an `if/else`-type condition.

Something along the lines of:

`if(!(old %in% colnames)){ # Checks if old columns names do not exist in existing column names 
    next # Skip to the next item in the list if TRUE
}`

The reason I would check it with a `! (NOT)` is because the very nature of `setnames()` checks to see if the names **do** exist; if so, it continues as normal because that is literally what it is designed for.

I have implemented `setnames()` as part of a large and complicated function (500 lines) that I have built and to be presented with an error after having spent quite some time waiting for it to execute is unhelpful.

A function like this would most certainly make things easier for me, at least, and I'm sure several others would agree.

Is this something that is being considered?
